builder/builder.go

```diff
@@ -199,15 +204,142 @@ func (b *Builder) startRegistryProxy(ctx context.Context, caps apicaps.CapSet) (
 		doneCh <- struct{}{}
 	}()
 
+	isDarwin := runtime.GOOS == "darwin"
+
+	if isDarwin {
+		darwinProxyPort, err := b.startDarwinProxyHelper(ctx, registryPort)
+		if err != nil {
+			b.opt.Console.Warnf("Failed to start registry proxy support container: %v", err)
+			err := b.stopDarwinProxyHelper(ctx) // Clean up dangling container on failure.
+			if err != nil {
+				b.opt.Console.Warnf("Failed to stop registry proxy support container: %v", err)
+			}
+			return nil, false
+		}
+		b.opt.Console.VerbosePrintf("Started support container for registry proxy on port %d", darwinProxyPort)
+		addr = fmt.Sprintf("localhost:%d", darwinProxyPort)
+	}
+
 	b.opt.LocalRegistryAddr = addr
 
-	return func() {
+	closeFn := func() {
+		if isDarwin {
+			err := b.stopDarwinProxyHelper(ctx)
+			if err != nil {
+				b.opt.Console.Warnf("Failed to stop registry proxy support container: %v", err)
+			}
+		}
 		p.Close()
 		select {
 		case <-ctx.Done():
 		case <-doneCh:
 		}
-	}, true
+	}
+
+	return closeFn, true
+}
+
+const darwinProxyContainerName = "darwin-registry-proxy"
+
+// startDarwinProxyHelper: Since Docker Desktop (Mac) containers run in a VM, a
+// special host name, host.docker.internal, is made available to access the
+// host's network. Docker can only pull insecurely from localhost, so we use a
+// socat container to proxy localhost:<port> request back out to the local
+// registry proxy created above.
+func (b *Builder) startDarwinProxyHelper(ctx context.Context, registryPort int) (int, error) {
+	infoMap, err := b.opt.ContainerFrontend.ContainerInfo(ctx, darwinProxyContainerName)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to check info")
+	}
+
+	var exists bool
+	if info, ok := infoMap[darwinProxyContainerName]; ok && info.Status != "missing" {
+		exists = true
+	}
+
+	if exists {
+		err := b.opt.ContainerFrontend.ContainerRemove(ctx, true, darwinProxyContainerName)
+		if err != nil {
+			return 0, errors.Wrap(err, "failed to remove existing container")
+		}
+	}
+
+	runCfg := containerutil.ContainerRun{
+		NameOrID: darwinProxyContainerName,
+		ImageRef: "alpine/socat:1.7.4.4",
```

Comment:
should this tag be override-able via a config value? I could see a case where a user has to pin to an older version of earthly; however something occurs to this specific tag and it gets removed from dockerhub. This would mean the user would have to pull this image from some other registry, or fake this tag locally to point to a different image.